src/collections.ts

```diff
@@ -0,0 +1,147 @@
+
+type Collections = typeof collections
+
+function filterNotNullish<T>(this: Array<T>): Array<NonNullable<T>> {
+  return this.filter((it): it is NonNullable<T> => it != undefined)
+}
+
+function extendArray<
+  KEY extends {
+    [P in keyof Collections]: Collections[P] extends (
+      first: any[],
+      ...args: any
+    ) => any
+      ? P
+      : never
+  }[keyof Collections]
+>(key: KEY): void {
+  extendProtoype(Array, apply(collections[key]), key)
+}
+
+extendProtoype(Array, filterNotNullish, "filterNotNullish")
+
+extendArray("associateBy")
+extendArray("associateWith")
+extendArray("chunk")
+extendArray("distinct")
+extendArray("distinctBy")
+extendArray("dropLastWhile")
+extendArray("dropWhile")
+extendArray("findLast")
+extendArray("findLastIndex")
+extendArray("findSingle")
+extendArray("firstNotNullishOf")
+extendArray("groupBy")
+extendArray("intersect")
+extendArray("mapNotNullish")
+extendArray("maxBy")
+extendArray("maxOf")
+extendArray("maxWith")
+extendArray("minBy")
+extendArray("minOf")
+extendArray("minWith")
+extendArray("partition")
+extendArray("permutations")
+extendArray("runningReduce")
+extendArray("sample")
+extendArray("slidingWindows")
+extendArray("sortBy")
+extendArray("sumOf")
+extendArray("takeLastWhile")
+extendArray("takeWhile")
+extendArray("takeWhile")
+extendArray("union")
+extendArray("unzip")
+extendArray("withoutAll")
+extendArray("zip")
+
+type PairSplit<T> = T extends [infer F, infer L] ? [F[], L[]] : never
+
+type StringArrayRecord<T, U> = T extends string ? Record<string, U> : never
+
+declare global {
+  interface Array<T> {
+    associateBy(selector: (el: T) => string): Record<string, T>
+    associateWith<U>(selector: (key: string) => U): StringArrayRecord<T, U>
+    chunk(size: number): T[][]
+    distinct(): T[]
+    distinctBy<D>(selector: (el: T) => D): T[]
+    dropLastWhile(predicate: (el: T) => boolean): T[]
+    dropWhile(predicate: (el: T) => boolean): T[]
+    filterNotNullish(): Array<NonNullable<T>>
+    findLast(predicate: (el: T) => boolean): T | undefined
+    findLastIndex(predicate: (el: T) => boolean): number | undefined
+    findSingle(predicate: (el: T) => boolean): T | undefined
+    firstNotNullishOf<O>(
+      selector: (item: T) => O | undefined | null
+    ): NonNullable<O> | undefined
+    groupBy(selector: (el: T) => string): Record<string, T[]>
+    intersect(...arrays: (readonly T[])[]): T[]
+    mapNotNullish<O>(transformer: (el: T) => O): NonNullable<O>[]
+    maxBy(
+      selector:
+        | ((el: T) => number)
+        | ((el: T) => string)
+        | ((el: T) => bigint)
+        | ((el: T) => Date)
+    ): T | undefined
+    maxBy(selector: (el: T) => string): T | undefined
+    maxBy(selector: (el: T) => bigint): T | undefined
+    maxBy(selector: (el: T) => number): T | undefined
+    maxBy(selector: (el: T) => Date): T | undefined
+    maxOf<S extends ((el: T) => number) | ((el: T) => bigint)>(
+      selector: S
+    ): ReturnType<S> | undefined
```

Comment:
You don't need that - looks like you coped overloads **and** their implementation signatures, while only the overloads are actually exposed. That goes for all overloaded functions here.